Use a more clever way to replace parts of a file

killercup opened the pull request:
Addresses #57

Btw, this is still only concerning itself with replacing multiple things in one file. #66 changes the CLI to group suggestions by file. Not sure if we want to make this part of the library.

killercup commented:
Works great on top of #66 after adding a single `?`

alexcrichton commented:
Nice! I think that merging the `replace-rs` repo into this one (if you're willing) would probably be a good idea if it's pretty rustfix-specific. Other than that mind also adding a few tests in the `cargo-fix/tests` folder for situations that have now been fixed?

alexcrichton commented:
Also you may want to rebase on master now as it turns out I forgot to include running of the new tests in the branch this is based off!

alexcrichton commented on Cargo.toml:
```diff
@@ -20,13 +20,16 @@ exclude = [
 serde = "1.0"
 serde_json = "1.0"
 serde_derive = "1.0"
+replace-rs = { git = "https://github.com/killercup/replace-rs" }
```
This can be removed now I think

alexcrichton commented on src/lib.rs:
```diff
@@ -208,16 +219,22 @@ pub fn apply_suggestion(file_content: &mut String, suggestion: &Replacement) ->
     new_content
```
Should this method be removed now?

killercup replied:
Yep. Breaking change! 🙌

alexcrichton commented on src/replace.rs:
```diff
@@ -0,0 +1,223 @@
+
+    proptest! {
+        #[test]
+        #[ignore]
```
How come these are `#[ignore]`? Do they take too long?

killercup replied:
They take a few seconds each, yeah. I don't really want to run them on each `cargo test`, but I'll add them to CI

killercup commented:
Moved it in as a module, and added tests from #77 which automagically work now.

I've also added another test for "overlapping" suggestions. The only one I could think of that I know exists in rustc is missing imports with multiple options. If you have an idea for other tests, let me know (or just add them :))

alexcrichton commented on cargo-fix/tests/all/smoke.rs:
```diff
@@ -87,3 +87,44 @@ fn tricky_ampersand() {
+
+    let stderr = "\
+[CHECKING] foo v0.1.0 (CWD)
+error: Cannot replace slice of data that was already replaced
```
Hm I'm not sure that this is the UI we're gonna want long-term, I think we probably want to just skip auto-fixing code and let warnings like these naturally bubble up to the user?

Could we add some logic to avoid attempting to apply suggestions that look like they're gonna fail? (or otherwise just keep track of suggestions that fail to apply)

killercup replied:
Yes that's why I was asking for a better test case :) I opened #78 for the intended user experience

alexcrichton replied:
Hm could we tweak this test to assert that the warning isn't updated here? I think it's probably most conservative to start out by not applying any suggestions when they're ambiguous about what to do. Anything we don't fix will end up being seen by the end user anyway

killercup replied:
Yeah. Do you want me to do this here or in another PR? (The behavior isn't changed here AFAIK)

alexcrichton replied:
Ah ok let's do that in a follow-up then

killercup commented on tests/everything.rs:
```diff
@@ -15,14 +15,15 @@ use std::collections::HashSet;
         "rustc", file,
         "--error-format=pretty-json", "-Zunstable-options", "--emit=metadata",
         "--crate-name=rustfix_test",
+        "-Zsuggestion-applicability",
```
Oh! Don't look! Spoilers!
